### PR TITLE
Optimize the number of db queries to build the main page.

### PR DESCRIPTION
**Pull request by vdbergh**
Ideas by peregrineshahin, vondele, xoto10.

See #1596, #1597, #1620.

### CODE REVIEW

**Comment by vdbergh** [comment]
Untested so likely will not work. Hence currently draft. I will test it later.

File: server/fishtest/rundb.py [hunk 1]
```diff
@@ -237,13 +237,13 @@ def new_run(
...
-        active_runs = self.runs.find(
-            {"finished": False, "tasks": {"$elemMatch": {"active": True}}},
-            sort=[("last_updated", DESCENDING)],
-        )
-        for run in active_runs:
+        #        active_runs = self.runs.find(
+        #            {"finished": False, "tasks": {"$elemMatch": {"active": True}}},
+        #            sort=[("last_updated", DESCENDING)],
+        #        )
+        for run in unfinishd_runs:
```
Comment (vondele) [inline server/fishtest/rundb.py:246]: typo

File: server/fishtest/rundb.py [hunk 2]
```diff
@@ -476,14 +476,18 @@ def get_unfinished_runs(self, username=None):
             unfinished_runs = self.runs.find(
                 {"finished": False}, sort=[("last_updated", DESCENDING)]
             )
+            # unfished_runs is a cursor, convert to list
```
Comment (vondele) [inline server/fishtest/rundb.py:479]: typo in comment

File: server/fishtest/rundb.py [hunk 3]
```diff
@@ -476,14 +476,18 @@ def get_unfinished_runs(self, username=None):
...
             if username:
                 unfinished_runs = [
                     r for r in unfinished_runs if r["args"].get("username") == username
                 ]
+            else:
+                unfinished_runs = list(unfinishd_runs)
```
Comment (vondele) [inline server/fishtest/rundb.py:485]: typo

File: server/fishtest/rundb.py [hunk 4]
```diff
@@ -237,13 +237,9 @@ def new_run(
...
 
-    def get_machines(self):
+    def get_machines(self, unfinished_runs):
         machines = []
-        active_runs = self.runs.find(
-            {"finished": False, "tasks": {"$elemMatch": {"active": True}}},
-            sort=[("last_updated", DESCENDING)],
-        )
-        for run in active_runs:
+        for run in unfinished_runs:
```
Comment (Sopel97) [inline server/fishtest/rundb.py:242]: missing filtering by active?

Comment (peregrineshahin) [reply server/fishtest/rundb.py:242]: yes Indeed. but this was xoto idea here https://github.com/glinscott/fishtest/pull/1597/
Which I don't like.

We reduced 3 dp calls by this.
We can afford filtering in code here.

Comment (Sopel97) [reply server/fishtest/rundb.py:242]: the query should definitely stay, was thinking about filtering in the code, as I wasn't sure it's equivalent without it

Comment (peregrineshahin) [reply server/fishtest/rundb.py:242]: No queries in this line, we have the data passed, so I definitely agree with you that, to make this equivalent, A filtering in code to active runs would be best and not at all expensive

Comment (vdbergh) [reply server/fishtest/rundb.py:242]: There is already filtering in code right (line 244)?

**Comment by vdbergh** [comment]
I did some quick testing and the PR appeared to work as intended, so I have undrafted it.

**Comment by vdbergh** [comment]
I suspect the most expensive operation is
```python
unfinished_runs = list(unfinished_runs)
```
which converts the cursor to a list (in other words, it retrieves the documents from the db).

**Comment by vdbergh** [comment]
I converted to draft again since there is still a bit of double work. Both `get_machines()` and `aggregate_unfinished_runs()` need to determine the active runs. I think we do not actually need `get_machines()`. We can incorporate it in `aggregate_unfinished_runs()` .

**Comment by vdbergh** [comment]
Undrafted again. Now `get_machines()` has been eliminated.

**Comment by ppigazzini** [comment]
> I suspect the most expensive operation is
> 
> ```python
> unfinished_runs = list(unfinished_runs)
> ```
> 
> which converts the cursor to a list (in other words, it retrieves the documents from the db).

We added the lock with https://github.com/glinscott/fishtest/pull/178 because https://www.mongodb.com/docs/manual/core/read-isolation-consistency-recency/#read-uncommitted-and-multiple-document-write
> Without isolating the multi-document write operations, MongoDB exhibits the following behavior:
> ...
> 3. Reads may miss matching documents that are updated during the course of the read operation.

Here an analysis of the MongoDB behavior. They solved their problem using a query with an index not modified during the concurrent write operations:
https://blog.meteor.com/mongodb-queries-dont-always-return-all-matching-documents-654b6594a827

Without that lock the `get_unfinished_run()` could return or the cursor or the generator expression filtering the cursor with the username, making a lazy query in the callers (at the moment all the callers are iterating only once). This will save surely some RAM and perhaps some CPU time.

**Comment by ppigazzini** [comment]
DEV updated, 258 workers connected (courtesy of the mongodb backup).

**Comment by vondele** [comment]
functionality seems OK at first sight, but I don't measure a speedup.

**Comment by ppigazzini** [comment]
DEV is slower than PROD.

**Comment by vondele** [comment]
I did some measurements earlier on dev.

**Comment by ppigazzini** [comment]
OK.

**Comment by vondele** [comment]
did shouldn't necessarily stop this PR, as I think this is a cleaner implementation. But it is a bit surprising. 
Maybe a cProfile on this would be valuable?

**Comment by ppigazzini** [comment]
DEV instrumented. Please keep in mind the DEV has still nginx configuration with 3 `pserve`.

**Comment by vondele** [comment]
OK, triggered it a couple of times. We should have the profile data.

**Comment by ppigazzini** [comment]
[cprofile_pr1621.zip](https://github.com/glinscott/fishtest/files/11299290/cprofile_pr1621.zip)

**Comment by vondele** [comment]
well, it does remove the calls as advertised:
```
      1    0.001    0.001    4.320    4.320 /home/usr00/fishtest/server/fishtest/views.py:1381(homepage_results)
   14/8    0.000    0.000    4.147    0.518 /home/usr00/fishtest/server/env/lib/python3.11/site-packages/pymongo/_csot.py:98(csot_wrapper)
      7    0.000    0.000    4.145    0.592 /home/usr00/fishtest/server/env/lib/python3.11/site-packages/pymongo/mongo_client.py:1418(_retryable_read)
    125    0.000    0.000    4.047    0.032 /home/usr00/fishtest/server/env/lib/python3.11/site-packages/pymongo/cursor.py:1244(next)
      9    0.000    0.000    4.047    0.450 /home/usr00/fishtest/server/env/lib/python3.11/site-packages/pymongo/cursor.py:1128(_refresh)
      6    0.000    0.000    4.046    0.674 /home/usr00/fishtest/server/env/lib/python3.11/site-packages/pymongo/cursor.py:1037(__send_message)
      6    0.000    0.000    4.046    0.674 /home/usr00/fishtest/server/env/lib/python3.11/site-packages/pymongo/mongo_client.py:1302(_run_operation)
      6    0.000    0.000    4.043    0.674 /home/usr00/fishtest/server/env/lib/python3.11/site-packages/pymongo/mongo_client.py:1324(_cmd)
      6    0.000    0.000    4.043    0.674 /home/usr00/fishtest/server/env/lib/python3.11/site-packages/pymongo/server.py:76(run_operation)
      8    0.000    0.000    3.845    0.481 /home/usr00/fishtest/server/env/lib/python3.11/site-packages/pymongo/message.py:1391(unpack_response)
      6    0.000    0.000    3.845    0.641 /home/usr00/fishtest/server/env/lib/python3.11/site-packages/pymongo/cursor.py:1117(_unpack_response)
      8    0.000    0.000    3.845    0.481 /home/usr00/fishtest/server/env/lib/python3.11/site-packages/bson/__init__.py:1184(_decode_all_selective)
      8    0.000    0.000    3.845    0.481 /home/usr00/fishtest/server/env/lib/python3.11/site-packages/bson/__init__.py:1093(decode_all)
      8    3.843    0.480    3.845    0.481 {built-in method bson._cbson._decode_all}
      2    0.000    0.000    2.869    1.434 /home/usr00/fishtest/server/fishtest/rundb.py:458(get_unfinished_runs)
      1    0.023    0.023    1.467    1.467 /home/usr00/fishtest/server/fishtest/rundb.py:473(aggregate_unfinished_runs)
      1    0.000    0.000    1.277    1.277 /home/usr00/fishtest/server/fishtest/views.py:1310(get_paginated_finished_runs)
      1    0.000    0.000    1.270    1.270 /home/usr00/fishtest/server/fishtest/rundb.py:536(get_finished_runs)
      1    0.000    0.000    1.170    1.170 /home/usr00/fishtest/server/fishtest/rundb.py:562(<listcomp>)
      8    0.000    0.000    0.267    0.033 /home/usr00/fishtest/server/env/lib/python3.11/site-packages/pymongo/network.py:205(receive_message)
     16    0.013    0.001    0.267    0.017 /home/usr00/fishtest/server/env/lib/python3.11/site-packages/pymongo/network.py:287(_receive_data_on_socket)
    227    0.252    0.001    0.252    0.001 {method 'recv_into' of '_socket.socket' objects}
      6    0.000    0.000    0.183    0.031 /home/usr00/fishtest/server/env/lib/python3.11/site-packages/pymongo/pool.py:813(receive_message)
     48    0.002    0.000    0.117    0.002 /home/usr00/fishtest/server/fishtest/util.py:346(remaining_hours)
      2    0.000    0.000    0.100    0.050 /home/usr00/fishtest/server/env/lib/python3.11/site-packages/pymongo/pool.py:707(command)
      2    0.000    0.000    0.100    0.050 /home/usr00/fishtest/server/env/lib/python3.11/site-packages/pymongo/network.py:40(command)
      1    0.000    0.000    0.100    0.100 /home/usr00/fishtest/server/env/lib/python3.11/site-packages/pymongo/collection.py:1759(count_documents)
      1    0.000    0.000    0.100    0.100 /home/usr00/fishtest/server/env/lib/python3.11/site-packages/pymongo/collection.py:1844(_retryable_non_cursor_read)
      1    0.000    0.000    0.099    0.099 /home/usr00/fishtest/server/env/lib/python3.11/site-packages/pymongo/collection.py:1836(_cmd)
      1    0.000    0.000    0.099    0.099 /home/usr00/fishtest/server/env/lib/python3.11/site-packages/pymongo/collection.py:1701(_aggregate_one_result)
      1    0.000    0.000    0.099    0.099 /home/usr00/fishtest/server/env/lib/python3.11/site-packages/pymongo/collection.py:250(_command)
      1    0.000    0.000    0.092    0.092 /home/usr00/fishtest/server/env/lib/python3.11/site-packages/pyramid/renderers.py:433(render_view)
      1    0.000    0.000    0.092    0.092 /home/usr00/fishtest/server/env/lib/python3.11/site-packages/pyramid/renderers.py:465(render_to_response)
      1    0.000    0.000    0.092    0.092 /home/usr00/fishtest/server/env/lib/python3.11/site-packages/pyramid/renderers.py:445(render)
      1    0.000    0.000    0.092    0.092 /home/usr00/fishtest/server/env/lib/python3.11/site-packages/pyramid_mako/__init__.py:129(__call__)
      1    0.000    0.000    0.092    0.092 /home/usr00/fishtest/server/env/lib/python3.11/site-packages/mako/template.py:441(render_unicode)
      1    0.000    0.000    0.092    0.092 /home/usr00/fishtest/server/env/lib/python3.11/site-packages/mako/runtime.py:860(_render)
      1    0.000    0.000    0.091    0.091 /home/usr00/fishtest/server/env/lib/python3.11/site-packages/mako/runtime.py:908(_render_context)
      1    0.000    0.000    0.091    0.091 /home/usr00/fishtest/server/env/lib/python3.11/site-packages/mako/runtime.py:923(_exec_template)

```

**Comment by vondele** [comment]
There are still two calls to get_unfinished_runs

**Comment by ppigazzini** [comment]
cProfile with master
[cprofile_master.zip](https://github.com/glinscott/fishtest/files/11299349/cprofile_master.zip)

**Comment by vondele** [comment]
```
cprofile_pr1621_6544:Apr 3.364
cprofile_pr1621_6544:Apr 3.668
cprofile_pr1621_6544:Apr 3.699
cprofile_pr1621_6544:Apr 4.280
cprofile_pr1621_6544:Apr 3.654
cprofile_pr1621_6544:Apr 4.320
cprofile_master_6544:Apr 4.944
cprofile_master_6544:Apr 7.551
```
so master is in fact a little slower, the effect is not very large.

**Comment by vondele** [comment]
looks like a mistake in the PR. Let me have a closer look.

File: server/fishtest/views.py [hunk 5]
```diff
@@ -1370,15 +1370,25 @@ def tests_user(request):
...
+        )[0]
     # page 2 and beyond only show finished test results
     return response
 
 
 def homepage_results(request):
     # Calculate games_per_minute from current machines
     games_per_minute = 0.0
-    machines = request.rundb.get_machines()
+    unfinished_runs = request.rundb.get_unfinished_runs()
```
Comment (vondele) [inline server/fishtest/views.py:1383]: spurious line: `unfinished_runs = request.rundb.get_unfinished_runs()`

**Comment by vondele** [comment]
line 1383 in the PR is spurious and can just be deleted. `unfinished_runs = ` is unused.

**Comment by ppigazzini** [comment]
I'll fix DEV by hand and profile.

**Comment by vondele** [comment]
fixed :-)

**Comment by ppigazzini** [comment]
[cprofile_pr1621fix.zip](https://github.com/glinscott/fishtest/files/11299463/cprofile_pr1621fix.zip)

**Comment by vondele** [comment]
correct:
```
       1    0.001    0.001    2.181    2.181 /home/usr00/fishtest/server/fishtest/views.py:1381(homepage_results)
    10/6    0.000    0.000    2.032    0.339 /home/usr00/fishtest/server/env/lib/python3.11/site-packages/pymongo/_csot.py:98(csot_wrapper)
       5    0.000    0.000    2.029    0.406 /home/usr00/fishtest/server/env/lib/python3.11/site-packages/pymongo/mongo_client.py:1418(_retryable_read)
      76    0.000    0.000    1.887    0.025 /home/usr00/fishtest/server/env/lib/python3.11/site-packages/pymongo/cursor.py:1244(next)
       6    0.000    0.000    1.886    0.314 /home/usr00/fishtest/server/env/lib/python3.11/site-packages/pymongo/cursor.py:1128(_refresh)
       4    0.000    0.000    1.886    0.472 /home/usr00/fishtest/server/env/lib/python3.11/site-packages/pymongo/cursor.py:1037(__send_message)
       4    0.000    0.000    1.886    0.471 /home/usr00/fishtest/server/env/lib/python3.11/site-packages/pymongo/mongo_client.py:1302(_run_operation)
       4    0.000    0.000    1.884    0.471 /home/usr00/fishtest/server/env/lib/python3.11/site-packages/pymongo/mongo_client.py:1324(_cmd)
       4    0.000    0.000    1.884    0.471 /home/usr00/fishtest/server/env/lib/python3.11/site-packages/pymongo/server.py:76(run_operation)
       6    0.000    0.000    1.761    0.293 /home/usr00/fishtest/server/env/lib/python3.11/site-packages/pymongo/message.py:1391(unpack_response)
       6    0.000    0.000    1.761    0.293 /home/usr00/fishtest/server/env/lib/python3.11/site-packages/bson/__init__.py:1184(_decode_all_selective)
       6    0.000    0.000    1.761    0.293 /home/usr00/fishtest/server/env/lib/python3.11/site-packages/bson/__init__.py:1093(decode_all)
       4    0.000    0.000    1.761    0.440 /home/usr00/fishtest/server/env/lib/python3.11/site-packages/pymongo/cursor.py:1117(_unpack_response)
       6    1.760    0.293    1.761    0.293 {built-in method bson._cbson._decode_all}
       1    0.000    0.000    1.187    1.187 /home/usr00/fishtest/server/fishtest/views.py:1310(get_paginated_finished_runs)
       1    0.000    0.000    1.180    1.180 /home/usr00/fishtest/server/fishtest/rundb.py:536(get_finished_runs)
       1    0.000    0.000    1.037    1.037 /home/usr00/fishtest/server/fishtest/rundb.py:562(<listcomp>)
       1    0.021    0.021    0.992    0.992 /home/usr00/fishtest/server/fishtest/rundb.py:473(aggregate_unfinished_runs)
       1    0.000    0.000    0.842    0.842 /home/usr00/fishtest/server/fishtest/rundb.py:458(get_unfinished_runs)
       6    0.000    0.000    0.241    0.040 /home/usr00/fishtest/server/env/lib/python3.11/site-packages/pymongo/network.py:205(receive_message)
      12    0.010    0.001    0.240    0.020 /home/usr00/fishtest/server/env/lib/python3.11/site-packages/pymongo/network.py:287(_receive_data_on_socket)
     474    0.229    0.000    0.229    0.000 {method 'recv_into' of '_socket.socket' objects}
       2    0.000    0.000    0.145    0.073 /home/usr00/fishtest/server/env/lib/python3.11/site-packages/pymongo/pool.py:707(command)
       2    0.000    0.000    0.145    0.073 /home/usr00/fishtest/server/env/lib/python3.11/site-packages/pymongo/network.py:40(command)
       1    0.000    0.000    0.143    0.143 /home/usr00/fishtest/server/env/lib/python3.11/site-packages/pymongo/collection.py:1759(count_documents)
       1    0.000    0.000    0.143    0.143 /home/usr00/fishtest/server/env/lib/python3.11/site-packages/pymongo/collection.py:1844(_retryable_non_cursor_read)
       1    0.000    0.000    0.143    0.143 /home/usr00/fishtest/server/env/lib/python3.11/site-packages/pymongo/collection.py:1836(_cmd)
       1    0.000    0.000    0.143    0.143 /home/usr00/fishtest/server/env/lib/python3.11/site-packages/pymongo/collection.py:1701(_aggregate_one_result)
       1    0.000    0.000    0.143    0.143 /home/usr00/fishtest/server/env/lib/python3.11/site-packages/pymongo/collection.py:250(_command)
       4    0.000    0.000    0.115    0.029 /home/usr00/fishtest/server/env/lib/python3.11/site-packages/pymongo/pool.py:813(receive_message)
      48    0.002    0.000    0.096    0.002 /home/usr00/fishtest/server/fishtest/util.py:346(remaining_hours)
      45    0.000    0.000    0.071    0.002 /home/usr00/fishtest/server/env/lib/python3.11/site-packages/scipy/stats/_distn_infrastructure.py:828(__call__)
      45    0.000    0.000    0.071    0.002 /home/usr00/fishtest/server/env/lib/python3.11/site-packages/scipy/stats/_distn_infrastructure.py:808(freeze)
      45    0.001    0.000    0.070    0.002 /home/usr00/fishtest/server/env/lib/python3.11/site-packages/scipy/stats/_distn_infrastructure.py:435(__init__)
      45    0.002    0.000    0.069    0.002 /home/usr00/fishtest/server/env/lib/python3.11/site-packages/scipy/stats/_distn_infrastructure.py:1900(__init__)
       1    0.000    0.000    0.064    0.064 /home/usr00/fishtest/server/env/lib/python3.11/site-packages/pyramid/renderers.py:433(render_view)
       1    0.000    0.000    0.064    0.064 /home/usr00/fishtest/server/env/lib/python3.11/site-packages/pyramid/renderers.py:465(render_to_response)
       1    0.000    0.000    0.063    0.063 /home/usr00/fishtest/server/env/lib/python3.11/site-packages/pyramid/renderers.py:445(render)
       1    0.000    0.000    0.063    0.063 /home/usr00/fishtest/server/env/lib/python3.11/site-packages/pyramid_mako/__init__.py:129(__call__)
```

**Comment by ppigazzini** [comment]
We can call it a day :)

**Comment by vondele** [comment]
admittedly, with just 3 minutes left, we had no choice.

**Comment by ppigazzini** [comment]
New day!
Returning cursor/generator expression
[cprofile_pr1621_cursor_6544.zip](https://github.com/glinscott/fishtest/files/11299498/cprofile_pr1621_cursor_6544.zip)

**Comment by vondele** [comment]
yeah, CPU time is similar, or maybe even a little slower (there is quite some variation in this data)

**Comment by ppigazzini** [comment]
With the noise it should show a very little CPU speedup, but the real gain is in the RAM (if the query returns many data, of course).

**Comment by vondele** [comment]
active tests can be quite some RAM indeed (say 80 active tests, and each the maximum of 16MB mongo array for the tasks).

**Comment by ppigazzini** [comment]
We need to triple check if we can drop the lock before returning the cursor/generator, though.

**Comment by vondele** [comment]
indeed, "disappearing" test would be annoying.

File: server/fishtest/views.py (hunk 5 above)
Comment (vdbergh) [reply server/fishtest/views.py:1383]: Sorry about that!

**Comment by vdbergh** [comment]
> New day!
> Returning cursor/generator expression (I'm expecting saving RAM, only little CPU. The CPU load should move from the callee to the caller).
> [cprofile_pr1621_cursor_6544.zip](https://github.com/glinscott/fishtest/files/11299498/cprofile_pr1621_cursor_6544.zip)

Unfinished_runs is used only once now so it is indeed ok to return the cursor.

**Comment by vdbergh** [comment]
> We need to triple check if we can drop the lock before returning the cursor/generator, though.
> gn

I think the lock can be dropped safely.

**Comment by vdbergh** [comment]
> > I suspect the most expensive operation is
> > ```python
> > unfinished_runs = list(unfinished_runs)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > which converts the cursor to a list (in other words, it retrieves the documents from the db).
> 
> We added the lock with #178 because https://www.mongodb.com/docs/manual/core/read-isolation-consistency-recency/#read-uncommitted-and-multiple-document-write
> 
> > Without isolating the multi-document write operations, MongoDB exhibits the following behavior:
> > ...
> > 3. Reads may miss matching documents that are updated during the course of the read operation.
> 
> Here an analysis of the MongoDB behavior. They solved their problem using a query with an index not modified during the concurrent write operations: https://blog.meteor.com/mongodb-queries-dont-always-return-all-matching-documents-654b6594a827
> 
> > Long story short…
> > 
> > * This issue doesn’t affect queries that don’t use an index, such as queries that just look up a document by ID.
> > * It doesn’t affect queries which explicitly do a single value equality match on all fields used in the index key.
> > * It [mostly](https://jira.mongodb.org/browse/SERVER-14766) doesn’t affect queries that use indexes whose fields are never modified after the document is originally inserted.
> > * But any other kind of MongoDB query can fail to include all the matching documents!
> 
> Without the need of that lock, the `get_unfinished_run()` could return or the cursor or the generator expression built filtering the cursor with the username, making a lazy query in the callers (at the moment all the callers are iterating only once). This will save surely some RAM and perhaps some little CPU time.


The original motivation for introducing the lock no longer applies since reading and writing are now done in different instances. The lock only prevents parallel invocation of get_unfinished_runs() by different threads, but there is already a semaphore doing the same thing, and moreover such parallel invocations should be safe anyway since there is no shared data.

If disappearing tests would be a problem then it would have been noticed already.

**Comment by vdbergh** [comment]
@ppigazzini I think I see where the issue comes from. The indexes are sorted by last_updated. This means they are constantly reordered. The solution should be to do the sorting in code.

**Comment by vdbergh** [comment]
I removed the lock in `get_unfinished_runs()` and now I also return the naked cursor (this last thing is a matter of taste, the advantage of returning a list was that the result is then reusable).

I also made a companion PR #1623 that I think will solve the disappearing runs issue.

**Comment by vdbergh** [comment]
> active tests can be quite some RAM indeed (say 80 active tests, and each the maximum of 16MB mongo array for the tasks).

That would be retrieving  > 1Gb from the DB just to show the main page. 

The current design of Fishtest with the tasks being part of the runs is horribly non-scalable. Perhaps we should start thinking about moving the tasks to a separate flat table to solve this issue once and for all.

**Comment by vondele** [comment]
> > active tests can be quite some RAM indeed (say 80 active tests, and each the maximum of 16MB mongo array for the tasks).
> 
> That would be retrieving > 1Gb from the DB just to show the main page.

which isn't too inconsistent with the bjon conversion taking seconds. For me, it would be interesting to have once a sample of the .json that corresponds to the active runs dictionary.

> 
> The current design of Fishtest with the tasks being part of the runs is horribly non-scalable. Perhaps we should start thinking about moving the tasks to a separate flat table to solve this issue once and for all.

certainly to be considered.

**Comment by ppigazzini** [comment]
DEV updated with the companion PR as well.
[cprofile_pr1621_pr1623.zip](https://github.com/glinscott/fishtest/files/11300557/cprofile_pr1621_pr1623.zip)

**Comment by vondele** [comment]
looks fastest so far.

**Comment by vdbergh** [comment]
> DEV updated with the companion PR as well. [cprofile_pr1621_pr1623.zip](https://github.com/glinscott/fishtest/files/11300557/cprofile_pr1621_pr1623.zip)

Did you regenerate the index?

**Comment by ppigazzini** [comment]
No :(
Doing now.

**Comment by ppigazzini** [comment]
profiling with pr1621 and pr1623 after index regeneration on `runs` collection.
[cprofile_pr1621_pr1623_reindexed.zip](https://github.com/glinscott/fishtest/files/11300633/cprofile_pr1621_pr1623_reindexed.zip)

**Comment by vondele** [comment]
```
       1    0.001    0.001    1.974    1.974 fishtest/views.py:1381(homepage_results)
    10/6    0.000    0.000    1.851    0.308 pymongo/_csot.py:98(csot_wrapper)
       5    0.000    0.000    1.850    0.370 pymongo/mongo_client.py:1418(_retryable_read)
      76    0.000    0.000    1.781    0.023 pymongo/cursor.py:1244(next)
       6    0.000    0.000    1.780    0.297 pymongo/cursor.py:1128(_refresh)
       4    0.000    0.000    1.780    0.445 pymongo/cursor.py:1037(__send_message)
       4    0.000    0.000    1.780    0.445 pymongo/mongo_client.py:1302(_run_operation)
       4    0.000    0.000    1.778    0.444 pymongo/mongo_client.py:1324(_cmd)
       4    0.000    0.000    1.778    0.444 pymongo/server.py:76(run_operation)
       6    0.000    0.000    1.700    0.283 pymongo/message.py:1391(unpack_response)
       6    0.000    0.000    1.700    0.283 bson/__init__.py:1184(_decode_all_selective)
       4    0.000    0.000    1.700    0.425 pymongo/cursor.py:1117(_unpack_response)
       6    0.000    0.000    1.700    0.283 bson/__init__.py:1093(decode_all)
       6    1.699    0.283    1.700    0.283 {built-in method bson._cbson._decode_all}
       1    0.000    0.000    1.009    1.009 fishtest/views.py:1310(get_paginated_finished_runs)
       1    0.000    0.000    1.003    1.003 fishtest/rundb.py:531(get_finished_runs)
       1    0.017    0.017    0.964    0.964 fishtest/rundb.py:468(aggregate_unfinished_runs)
       1    0.000    0.000    0.932    0.932 fishtest/rundb.py:557(<listcomp>)
       6    0.000    0.000    0.146    0.024 pymongo/network.py:205(receive_message)
      12    0.008    0.001    0.146    0.012 pymongo/network.py:287(_receive_data_on_socket)
     161    0.137    0.001    0.137    0.001 {method 'recv_into' of '_socket.socket' objects}
       4    0.000    0.000    0.076    0.019 pymongo/pool.py:813(receive_message)
      48    0.001    0.000    0.074    0.002 fishtest/util.py:346(remaining_hours)
       1    0.000    0.000    0.070    0.070 pymongo/collection.py:1759(count_documents)
```

**Comment by vdbergh** [comment]
I did not expect the new index to help (or hinder) query speed. However it should help with the disappearing run issue.

**Comment by vdbergh** [comment]
The new index may also make the main pserve a bit more efficient since now it only has to be touched when a run is created, or when a run finishes. These events are much rarer than finishing tasks.

**Review by ppigazzini — APPROVED** [review]

**Comment by ppigazzini** [comment]
PROD updated (with the companion PR, index regenerated for `runs` collection).
Thnak you @vdbergh (and @peregrineshahin , @vondele) :)

Current indexes on `runs`:
```yaml
{ '_id_': {'key': [('_id', 1)], 'v': 2},
  'finished_green_runs': { 'key': [('finished', 1), ('is_green', -1), ('last_updated', -1)],
                           'partialFilterExpression': SON([('finished', True), ('is_green', True)]),
                           'v': 2},
  'finished_ltc_runs': { 'key': [('finished', 1), ('last_updated', -1), ('tc_base', -1)],
                         'partialFilterExpression': SON([('finished', True), ('tc_base', SON([('$gte', 40)]))]),
                         'v': 2},
  'finished_runs': { 'key': [('finished', 1), ('last_updated', -1)],
                     'partialFilterExpression': SON([('finished', True)]),
                     'v': 2},
  'finished_user_runs': { 'key': [('args.username', -1), ('finished', 1), ('last_updated', -1)],
                          'partialFilterExpression': SON([('finished', True)]),
                          'v': 2},
  'finished_yellow_runs': { 'key': [('finished', 1), ('is_yellow', -1), ('last_updated', -1)],
                            'partialFilterExpression': SON([('finished', True), ('is_yellow', True)]),
                            'v': 2},
  'unfinished_runs': { 'key': [('finished', 1)],
                       'partialFilterExpression': SON([('finished', False)]),
                       'v': 2},
  'user_runs': {'key': [('args.username', -1), ('last_updated', -1)], 'v': 2}}

Current indexes on deltas:
{'_id_': {'key': [('_id', 1)], 'v': 2}}

Current indexes on nns:
{'_id_': {'key': [('_id', 1)], 'v': 2}, 'name_-1': {'key': [('name', -1)], 'v': 2}}

Current indexes on flag_cache:
{ '_id_': {'key': [('_id', 1)], 'v': 2},
  'geoip_checked_at_1': {'expireAfterSeconds': 604800, 'key': [('geoip_checked_at', 1)], 'v': 2},
  'ip_1': {'key': [('ip', 1)], 'v': 2}}
```

**Comment by ppigazzini** [comment]
cProfile of "tests" route for PROD after the update.

[cprofile_prod_6544.zip](https://github.com/glinscott/fishtest/files/11301089/cprofile_prod_6544.zip)

**Comment by vondele** [comment]
nice:
```
ncalls  tottime  percall  cumtime  percall filename:lineno(function)
     1    0.000    0.000    1.176    1.176 pyramid/viewderivers.py:426(rendered_view)
     1    0.000    0.000    1.130    1.130 pyramid/viewderivers.py:139(_requestonly_view)
     1    0.063    0.063    1.130    1.130 fishtest/views.py:1423(tests)
     1    0.001    0.001    1.064    1.064 fishtest/views.py:1380(homepage_results)
   8/5    0.000    0.000    0.973    0.195 pymongo/_csot.py:98(csot_wrapper)
     4    0.000    0.000    0.971    0.243 pymongo/mongo_client.py:1418(_retryable_read)
    73    0.000    0.000    0.893    0.012 pymongo/cursor.py:1244(next)
     5    0.000    0.000    0.893    0.179 pymongo/cursor.py:1128(_refresh)
     3    0.000    0.000    0.893    0.298 pymongo/cursor.py:1037(__send_message)
     3    0.000    0.000    0.893    0.298 pymongo/mongo_client.py:1302(_run_operation)
     3    0.000    0.000    0.891    0.297 pymongo/mongo_client.py:1324(_cmd)
     3    0.000    0.000    0.891    0.297 pymongo/server.py:76(run_operation)
     5    0.000    0.000    0.842    0.168 pymongo/message.py:1391(unpack_response)
     5    0.000    0.000    0.842    0.168 bson/__init__.py:1184(_decode_all_selective)
     5    0.000    0.000    0.842    0.168 bson/__init__.py:1093(decode_all)
     3    0.000    0.000    0.842    0.281 pymongo/cursor.py:1117(_unpack_response)
     5    0.842    0.168    0.842    0.168 {built-in method bson._cbson._decode_all}
     1    0.012    0.012    0.614    0.614 fishtest/rundb.py:468(aggregate_unfinished_runs)
     1    0.000    0.000    0.447    0.447 fishtest/views.py:1309(get_paginated_finished_runs)
     1    0.000    0.000    0.442    0.442 fishtest/rundb.py:531(get_finished_runs)
     1    0.000    0.000    0.363    0.363 fishtest/rundb.py:557(<listcomp>)
     5    0.000    0.000    0.127    0.025 pymongo/network.py:205(receive_message)
    10    0.006    0.001    0.127    0.013 pymongo/network.py:287(_receive_data_on_socket)
   171    0.120    0.001    0.120    0.001 {method 'recv_into' of '_socket.socket' objects}
     2    0.000    0.000    0.079    0.040 pymongo/pool.py:707(command)
     1    0.000    0.000    0.079    0.079 pymongo/collection.py:1759(count_documents)
     1    0.000    0.000    0.079    0.079 pymongo/collection.py:1844(_retryable_non_cursor_read)
     2    0.000    0.000    0.079    0.040 pymongo/network.py:40(command)
     1    0.000    0.000    0.079    0.079 pymongo/collection.py:1836(_cmd)
     1    0.000    0.000    0.079    0.079 pymongo/collection.py:1701(_aggregate_one_result)
     1    0.000    0.000    0.079    0.079 pymongo/collection.py:250(_command)
    45    0.001    0.000    0.057    0.001 fishtest/util.py:346(remaining_hours)
     3    0.000    0.000    0.048    0.016 pymongo/pool.py:813(receive_message)
     1    0.000    0.000    0.046    0.046 pyramid/renderers.py:433(render_view)
     1    0.000    0.000    0.046    0.046 pyramid/renderers.py:465(render_to_response)
     1    0.000    0.000    0.046    0.046 pyramid/renderers.py:445(render)
     1    0.000    0.000    0.046    0.046 pyramid_mako/__init__.py:129(__call__)
     1    0.000    0.000    0.046    0.046 mako/template.py:441(render_unicode)
     1    0.000    0.000    0.046    0.046 mako/runtime.py:860(_render)
     1    0.000    0.000    0.045    0.045 mako/runtime.py:908(_render_context)
     1    0.000    0.000    0.045    0.045 mako/runtime.py:923(_exec_template)
     1    0.000    0.000    0.045    0.045 base_mak:15(render_body)
```

**Comment by vondele** [comment]
prod seems to have different nice and cpulimit settings for the crontab processes after the update.

**Comment by ppigazzini** [comment]
yes, the ghostadmin struck silently...

```bash
$ crontab -l
VENV=/home/fishtest/fishtest/server/env
UPATH=/home/fishtest/fishtest/server/utils

# Backup mongodb database every 6 hours and upload to s3
3 */6 * * * /usr/bin/nice -n 10 /usr/bin/cpulimit -l 50 -f -m -- sh ${UPATH}/backup.sh

# Update the users table
1,16,31,46 * * * * /usr/bin/nice -n 10 /usr/bin/cpulimit -l 10 -f -m -- ${VENV}/bin/python3 ${UPATH}/delta_update_users.py

# Purge old pgn files
33 3 * * * /usr/bin/nice -n 10 /usr/bin/cpulimit -l 10 -f -m -- ${VENV}/bin/python3 ${UPATH}/purge_pgn.py

# Clean up old mail (more than 9 days old)
32 5 * * * /usr/bin/nice -n 10 screen -D -m mutt -e 'push D~d>9d<enter>qy<enter>'
```

**Comment by vondele** [comment]
looks like this merge broke `server/utils/delta_update_users.py` as it relies on `get_machines()`.

**Comment by ppigazzini** [comment]
Fixed on the fly.

```diff
-    machines = rundb.get_machines()
+    __runs, machines, __pending_hours, __cores, __nps = rundb.aggregate_unfinished_runs()
```

**Pull request merged**